PR: https://github.com/odoo/odoo/pull/245216

From: cd6325d7476b20d9d0a3d298306990dd1cf7ef77
From: Nisarg (nipl)
Date: 2026-02-12 18:58:14

Breaking data model changes scores: del:2.4 + add:0, change matches:
-    timesheet_invoice_id = fields.Many2one('account.move', string="Invoice", readonly=True, copy=False, help="Invoice created from the timesheet", index='btree_not_null')
-    so_line = fields.Many2one(compute="_compute_so_line", store=True, readonly=False,
-    order_id = fields.Many2one(related='so_line.order_id', store=True, readonly=True, index=True)
+    so_line = fields.Many2one(

Total Changes: 213

[IMP] sale, _*: add Services & Materials upselling view

_*= sale_project, sale_timesheet

Added a new *Services & Materials* view to allow users to create additional delivered
flat fees service or consumable product entries directly in a sale order.

This feature allows:
* Creating analytic lines from a sale order for extra services and materials.
* Automatically creating or updating related sale order lines from those
  analytic lines.
* Handling SO line creation based on product expense policy:
   * `cost` → create a new SO line.
   * `sales_price` → update an existing line if available.
* Updating delivered quantities when analytic lines are changed.
* Reverting effect/deleting SO line when analytic lines are deleted.

All analytic lines created from this flow are assigned to the analytic plan
`sale.analytic_plan_sale_orders`, which ensures an analytic account is created
and set on the sale order.

Products with reinvoicing policy `no` will only be eligible for manual delivery.

If a product has reinvoicing policy anything other than `no` it will be eligible
for upselling from our new view, and delivered quantity will be aggregated
from AALs.

If a project is selected on the sale order, the analytic account is deduced from
that project instead.

Added a stat button on invoices to easily view analytic lines linked to that
invoice.

Technical changes:

* Field `timesheet_invoice_id` is renamed to `reinvoice_move_id` as it's not
  timesheet specific anymore.
* Made `product_uom_id` field on AAL compute to make sure we get expected
  `uom_id` on AAL.
* Changed field `order_id` from `related` to `compute` since we need it to work
  as normal m2o in sale.
* Centralized analytic line reinvoicing logic in the sale module.
* Improved `_get_analytic_lines_to_link` to support timesheet-based
  subscriptions with date filtering.
* `qty_delivered` will now be sum of upsell AALs and stock moves(for
   consu products) and recorded timesheet hours(for timesheets).

task-5490517

closes odoo/odoo#245216

Related: odoo/enterprise#105158
Related: odoo/upgrade#9366
Signed-off-by: Victor Feyens (vfe) <vfe@odoo.com>

================================= pseudo patch: =================================

--- a/addons/sale_timesheet/models/account_move.py
+++ b/addons/sale_timesheet/models/account_move.py
@@ -9,7 +9,15 @@ from odoo.fields import Domain
 class AccountMove(models.Model):
     _inherit = "account.move"
 
-    timesheet_ids = fields.One2many('account.analytic.line', 'timesheet_invoice_id', string='Timesheets', readonly=True, copy=False, export_string_translation=False)
+    timesheet_ids = fields.One2many(
+        'account.analytic.line',
+        'reinvoice_move_id',
+        string='Timesheets',
+        domain=[('project_id', '!=', False)],
+        readonly=True,
+        copy=False,
+        export_string_translation=False,
+    )
     timesheet_count = fields.Integer("Number of timesheets", compute='_compute_timesheet_count', compute_sudo=True, export_string_translation=False)
     timesheet_encode_uom_id = fields.Many2one('uom.uom', related='company_id.timesheet_encode_uom_id', export_string_translation=False)
     timesheet_total_duration = fields.Integer("Timesheet Total Duration",
@@ -22,8 +30,8 @@ class AccountMove(models.Model):
             self.timesheet_total_duration = 0
             return
         group_data = self.env['account.analytic.line']._read_group([
-            ('timesheet_invoice_id', 'in', self.ids)
-        ], ['timesheet_invoice_id'], ['unit_amount:sum'])
+            ('reinvoice_move_id', 'in', self.ids), ('project_id', '!=', False),
+        ], ['reinvoice_move_id'], ['unit_amount:sum'])
         timesheet_unit_amount_dict = defaultdict(float)
         timesheet_unit_amount_dict.update({timesheet_invoice.id: amount for timesheet_invoice, amount in group_data})
         for invoice in self:
@@ -36,10 +44,17 @@ class AccountMove(models.Model):
 
     @api.depends('timesheet_ids')
     def _compute_timesheet_count(self):
-        timesheet_data = self.env['account.analytic.line']._read_group([('timesheet_invoice_id', 'in', self.ids)], ['timesheet_invoice_id'], ['__count'])
-        mapped_data = {timesheet_invoice.id: count for timesheet_invoice, count in timesheet_data}
+        timesheet_data = self.env['account.analytic.line']._read_group(
+            [
+                ('reinvoice_move_id', 'in', self.ids),
+                ('project_id', '!=', False),
+            ],
+            ['reinvoice_move_id'],
+            ['__count'],
+        )
+        mapped_data = dict(timesheet_data)
         for invoice in self:
-            invoice.timesheet_count = mapped_data.get(invoice.id, 0)
+            invoice.timesheet_count = mapped_data.get(invoice, 0)
 
     def action_view_timesheet(self):
         self.ensure_one()
@@ -66,40 +81,31 @@ class AccountMove(models.Model):
             }
         }
 
-    def _link_timesheets_to_invoice(self, start_date=None, end_date=None):
-        """ Search timesheets from given period and link this timesheets to the invoice
-
-            When we create an invoice from a sale order, we need to
-            link the timesheets in this sale order to the invoice.
-            Then, we can know which timesheets are invoiced in the sale order.
-            :param start_date: the start date of the period
-            :param end_date: the end date of the period
-        """
-        for line in self.filtered(lambda i: i.move_type == 'out_invoice' and i.state == 'draft').invoice_line_ids:
-            sale_line_delivery = line.sale_line_ids.filtered(lambda sol: sol.product_id.invoice_policy == 'delivery' and sol.product_id.service_type == 'timesheet')
-            if not start_date and not end_date:
-                start_date, end_date = self._get_range_dates(sale_line_delivery.order_id)
-            if sale_line_delivery:
-                domain = Domain(line._timesheet_domain_get_invoiced_lines(sale_line_delivery))
-                if start_date:
-                    domain &= Domain('date', '>=', start_date)
-                if end_date:
-                    domain &= Domain('date', '<=', end_date)
-                timesheets = self.env['account.analytic.line'].sudo().search(domain)
-                timesheets.write({'timesheet_invoice_id': line.move_id.id})
-
-    def _get_range_dates(self, order):
-        # A method that can be overridden
-        # to set the start and end dates according to order values
-        return None, None
-
     def action_post(self):
         result = super().action_post()
         credit_notes = self.filtered(lambda move: move.move_type == 'out_refund' and move.reversed_entry_id)
         timesheets_sudo = self.env['account.analytic.line'].sudo().search([
-            ('timesheet_invoice_id', 'in', credit_notes.reversed_entry_id.ids),
+            ('reinvoice_move_id', 'in', credit_notes.reversed_entry_id.ids),
             ('so_line', 'in', credit_notes.invoice_line_ids.sale_line_ids.ids),
             ('project_id', '!=', False),
         ])
-        timesheets_sudo.write({'timesheet_invoice_id': False})
+        timesheets_sudo.write({'reinvoice_move_id': False})
         return result
+
+    def _domain_services_analytic_line(self):
+        """Make sure we see only services analytic line that are not timesheets."""
+        return super()._domain_services_analytic_line() & Domain('project_id', '=', False)
+
+    @api.model
+    def _analytic_line_domain_get_invoiced_lines(self, so_lines):
+        domain = super()._analytic_line_domain_get_invoiced_lines(so_lines)
+
+        start_date = self.env.context.get('timesheet_start_date', False)
+        end_date = self.env.context.get('timesheet_end_date', False)
+
+        if start_date:
+            domain &= Domain('date', '>=', start_date)
+        if end_date:
+            domain &= Domain('date', '<=', end_date)
+
+        return domain

--- a/addons/sale_timesheet/models/account_move_line.py
+++ b/addons/sale_timesheet/models/account_move_line.py
@@ -2,29 +2,12 @@
 
 from collections import defaultdict
 
-from odoo import api, models
+from odoo import models
 
 
 class AccountMoveLine(models.Model):
     _inherit = 'account.move.line'
 
-    @api.model
-    def _timesheet_domain_get_invoiced_lines(self, sale_line_delivery):
-        """ Get the domain for the timesheet to link to the created invoice
-            :param sale_line_delivery: recordset of sale.order.line to invoice
-            :return a normalized domain
-        """
-        return [
-            ('so_line', 'in', sale_line_delivery.ids),
-            ('project_id', '!=', False),
-            '|', '|',
-                ('timesheet_invoice_id', '=', False),
-                '&',
-                    ('timesheet_invoice_id.state', '=', 'cancel'),
-                    ('timesheet_invoice_id.payment_state', '!=', 'invoicing_legacy'),
-                ('timesheet_invoice_id.payment_state', '=', 'reversed')
-        ]
-
     def unlink(self):
         move_line_read_group = self.env['account.move.line'].search_read([
             ('move_id.move_type', '=', 'out_invoice'),
@@ -39,10 +22,10 @@ class AccountMoveLine(models.Model):
             sale_line_ids_per_move[move_line['move_id'][0]] += self.env['sale.order.line'].browse(move_line['sale_line_ids'])
 
         timesheet_read_group = self.sudo().env['account.analytic.line']._read_group([
-            ('timesheet_invoice_id.move_type', '=', 'out_invoice'),
-            ('timesheet_invoice_id.state', '=', 'draft'),
-            ('timesheet_invoice_id', 'in', self.move_id.ids)],
-            ['timesheet_invoice_id', 'so_line'],
+            ('reinvoice_move_id.move_type', '=', 'out_invoice'),
+            ('reinvoice_move_id.state', '=', 'draft'),
+            ('reinvoice_move_id', 'in', self.move_id.ids)],
+            ['reinvoice_move_id', 'so_line'],
             ['id:array_agg'])
 
         timesheet_ids = []
@@ -50,5 +33,5 @@ class AccountMoveLine(models.Model):
             if so_line.id in sale_line_ids_per_move[timesheet_invoice.id].ids:
                 timesheet_ids += ids
 
-        self.sudo().env['account.analytic.line'].browse(timesheet_ids).write({'timesheet_invoice_id': False})
+        self.sudo().env['account.analytic.line'].browse(timesheet_ids).write({'reinvoice_move_id': False})
         return super().unlink()

--- a/addons/sale_timesheet/models/account_move_reversal.py
+++ b/addons/sale_timesheet/models/account_move_reversal.py
@@ -8,8 +8,8 @@ class AccountMoveReversal(models.TransientModel):
         if is_modify:
             moves = self.move_ids.filtered(lambda m: m.move_type == 'out_invoice')
             timesheets_sudo = self.env['account.analytic.line'].sudo().search([
-                ('timesheet_invoice_id', 'in', moves.ids),
+                ('reinvoice_move_id', 'in', moves.ids),
             ])
             if timesheets_sudo:
-                timesheets_sudo.write({'timesheet_invoice_id': False})
+                timesheets_sudo.write({'reinvoice_move_id': False})
         return super().reverse_moves(is_modify)

--- a/addons/sale_timesheet/models/hr_timesheet.py
+++ b/addons/sale_timesheet/models/hr_timesheet.py
@@ -1,6 +1,6 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo.exceptions import UserError, ValidationError
+from odoo.exceptions import ValidationError
 
 from odoo import api, fields, models, _
 from odoo.fields import Domain
@@ -23,24 +23,24 @@ class AccountAnalyticLine(models.Model):
     _inherit = 'account.analytic.line'
 
     def _domain_so_line(self):
-        domain = Domain.AND([
+        domain = super()._domain_so_line()
+
+        return Domain.AND([
+            domain,
             self.env['sale.order.line']._sellable_lines_domain(),
             self.env['sale.order.line']._domain_sale_line_service(),
             [
                 ('order_partner_id.commercial_partner_id', '=', unquote('commercial_partner_id')),
             ],
         ])
-        return str(domain)
 
     timesheet_invoice_type = fields.Selection(TIMESHEET_INVOICE_TYPES, string="Billable Type",
             compute='_compute_timesheet_invoice_type', compute_sudo=True, store=True, readonly=True)
     commercial_partner_id = fields.Many2one('res.partner', compute="_compute_commercial_partner")
-    timesheet_invoice_id = fields.Many2one('account.move', string="Invoice", readonly=True, copy=False, help="Invoice created from the timesheet", index='btree_not_null')
-    so_line = fields.Many2one(compute="_compute_so_line", store=True, readonly=False,
-        domain=_domain_so_line, falsy_value_label="Non-billable",
-        help="Sales order item to which the time spent will be added in order to be invoiced to your customer. Remove the sales order item for the timesheet entry to be non-billable.")
-    # we needed to store it only in order to be able to groupby in the portal
-    order_id = fields.Many2one(related='so_line.order_id', store=True, readonly=True, index=True)
+    so_line = fields.Many2one(
+        falsy_value_label="Non-billable",
+        help="Sales order item to which the time spent will be added in order to be invoiced to your customer. Remove the sales order item for the timesheet entry to be non-billable."
+    )
     is_so_line_edited = fields.Boolean("Is Sales Order Item Manually Edited")
     allow_billable = fields.Boolean(related="project_id.allow_billable")
     sale_order_state = fields.Selection(related='order_id.state')
@@ -78,14 +78,23 @@ class AccountAnalyticLine(models.Model):
 
     @api.depends('task_id.sale_line_id', 'project_id.sale_line_id', 'employee_id', 'project_id.allow_billable')
     def _compute_so_line(self):
-        for timesheet in self.filtered(lambda t: not t.is_so_line_edited and t._is_not_billed()):  # Get only the timesheets are not yet invoiced
+        super()._compute_so_line()
+        # Get only the timesheets that are not yet invoiced
+        for timesheet in self.filtered(lambda t: t.project_id and not t.is_so_line_edited and t._is_not_billed()):
             timesheet.so_line = timesheet.project_id.allow_billable and timesheet._timesheet_determine_sale_line()
 
-    @api.depends('timesheet_invoice_id.state')
+    @api.depends('so_line')
+    def _compute_order_id(self):
+        super()._compute_order_id()
+        # compute only for timesheets
+        for timesheet in self.filtered('project_id'):
+            timesheet.order_id = timesheet.so_line.order_id
+
+    @api.depends('reinvoice_move_id.state')
     def _compute_partner_id(self):
         super(AccountAnalyticLine, self.filtered(lambda t: t._is_not_billed()))._compute_partner_id()
 
-    @api.depends('timesheet_invoice_id.state')
+    @api.depends('reinvoice_move_id.state')
     def _compute_project_id(self):
         super(AccountAnalyticLine, self.filtered(lambda t: t._is_not_billed()))._compute_project_id()
 
@@ -94,17 +103,23 @@ class AccountAnalyticLine(models.Model):
 
     def _is_not_billed(self):
         self.ensure_one()
-        return not self.timesheet_invoice_id or (self.timesheet_invoice_id.state == 'cancel' and self.timesheet_invoice_id.payment_state != 'invoicing_legacy')
+        return not self.reinvoice_move_id or (self.reinvoice_move_id.state == 'cancel' and self.reinvoice_move_id.payment_state != 'invoicing_legacy')
 
     def _check_timesheet_can_be_billed(self):
         return self.so_line in self.project_id.mapped('sale_line_employee_ids.sale_line_id') | self.task_id.sale_line_id | self.project_id.sale_line_id
 
-    def _check_can_write(self, values):
-        # prevent to update invoiced timesheets if one line is of type delivery
-        if self.sudo().filtered(lambda aal: aal.so_line.product_id.invoice_policy == "delivery") and self.filtered(lambda t: t.timesheet_invoice_id and t.timesheet_invoice_id.state != 'cancel'):
-            if any(field_name in values for field_name in ['unit_amount', 'employee_id', 'project_id', 'task_id', 'so_line', 'date']):
-                raise UserError(_('You cannot modify timesheets that are already invoiced.'))
-        return super()._check_can_write(values)
+    def _restricted_fields_when_invoiced(self):
+        return super()._restricted_fields_when_invoiced() + ['employee_id', 'project_id', 'task_id']
+
+    def _get_invoiced_line_write_error(self):
+        if self.project_id:
+            return self.env._("You cannot modify timsheets that are already invoiced.")
+        return super()._get_invoiced_line_write_error()
+
+    def _get_invoiced_line_delete_error(self):
+        if any(timesheet.project_id for timesheet in self):
+            return self.env._("You cannot remove timsheets that are already invoiced.")
+        return super()._get_invoiced_line_delete_error()
 
     def _timesheet_determine_sale_line(self):
         """ Deduce the SO line associated to the timesheet line:
@@ -151,24 +166,19 @@ class AccountAnalyticLine(models.Model):
         return [
             '|',
             '&',
-            ('timesheet_invoice_id', 'in', invoice_ids.ids),
+            ('reinvoice_move_id', 'in', invoice_ids.ids),
             # TODO : Master: Check if non_billable should be removed ?
             ('timesheet_invoice_type', 'in', ['billable_time', 'non_billable']),
             '&',
             ('timesheet_invoice_type', '=', 'billable_fixed'),
                 '&',
                 ('so_line', 'in', order_lines_ids.ids),
-                ('timesheet_invoice_id', '=', False),
+                ('reinvoice_move_id', '=', False),
         ]
 
     def _get_timesheets_to_merge(self):
         res = super()._get_timesheets_to_merge()
-        return res.filtered(lambda l: not l.timesheet_invoice_id or l.timesheet_invoice_id.state != 'posted')
-
-    @api.ondelete(at_uninstall=False)
-    def _unlink_except_invoiced(self):
-        if any(line.timesheet_invoice_id and line.timesheet_invoice_id.state == 'posted' for line in self):
-            raise UserError(_('You cannot remove a timesheet that has already been invoiced.'))
+        return res.filtered(lambda l: not l.reinvoice_move_id or l.reinvoice_move_id.state != 'posted')
 
     def _get_employee_mapping_entry(self):
         self.ensure_one()
@@ -200,7 +210,7 @@ class AccountAnalyticLine(models.Model):
             'res_model': 'account.move',
             'views': [[False, 'form']],
             'context': {'create': False},
-            'res_id': self.timesheet_invoice_id.id,
+            'res_id': self.reinvoice_move_id.id,
         }
 
     def _timesheet_convert_sol_uom(self, sol, to_unit):

--- a/addons/sale_timesheet/models/project_project.py
+++ b/addons/sale_timesheet/models/project_project.py
@@ -411,7 +411,7 @@ class ProjectProject(models.Model):
             return profitability_items
         aa_line_read_group = self.env['account.analytic.line'].sudo()._read_group(
             self.sudo()._get_profitability_aal_domain(),
-            ['timesheet_invoice_type', 'timesheet_invoice_id', 'currency_id', 'category'],
+            ['timesheet_invoice_type', 'reinvoice_move_id', 'currency_id', 'category'],
             ['amount:sum', 'id:array_agg'],
         )
         can_see_timesheets = with_action and len(self) == 1 and self.env.user.has_group('hr_timesheet.group_hr_timesheet_approver')

--- a/addons/sale_timesheet/models/sale_order.py
+++ b/addons/sale_timesheet/models/sale_order.py
@@ -156,10 +156,6 @@ class SaleOrder(models.Model):
                 line.has_displayed_warning_upsell = False
 
     def _create_invoices(self, grouped=False, final=False, date=None):
-        """Link timesheets to the created invoices. Date interval is injected in the
-        context in sale_make_invoice_advance_inv wizard.
-        """
         moves = super()._create_invoices(grouped=grouped, final=final, date=date)
-        moves._link_timesheets_to_invoice(self.env.context.get("timesheet_start_date"), self.env.context.get("timesheet_end_date"))
         self._reset_has_displayed_warning_upsell_order_lines()
         return moves

--- a/addons/sale_timesheet/models/sale_order_line.py
+++ b/addons/sale_timesheet/models/sale_order_line.py
@@ -62,14 +62,6 @@ class SaleOrderLine(models.Model):
                 remaining_hours = line.product_uom_id._compute_quantity(qty_left, uom_hour, round=False)
             line.remaining_hours = remaining_hours
 
-    @api.depends('product_id')
-    def _compute_qty_delivered_method(self):
-        """ Sale Timesheet module compute delivered qty for product [('type', 'in', ['service']), ('service_type', '=', 'timesheet')] """
-        super()._compute_qty_delivered_method()
-        for line in self:
-            if not line.is_expense and line.product_id.type == 'service' and line.product_id.service_type == 'timesheet':
-                line.qty_delivered_method = 'timesheet'
-
     @api.depends('analytic_line_ids.project_id', 'project_id.pricing_type')
     def _compute_qty_delivered(self):
         super()._compute_qty_delivered()
@@ -80,7 +72,7 @@ class SaleOrderLine(models.Model):
         domain = lines_by_timesheet._timesheet_compute_delivered_quantity_domain()
         mapping = lines_by_timesheet.sudo()._get_delivered_quantity_by_analytic(domain)
         for line in lines_by_timesheet:
-            delivered_qties[line] = mapping.get(line.id or line._origin.id, 0.0)
+            delivered_qties[line] += mapping.get(line.id or line._origin.id, 0.0)
         return delivered_qties
 
     def _timesheet_compute_delivered_quantity_domain(self):
@@ -90,6 +82,13 @@ class SaleOrderLine(models.Model):
             domain += [('date', '<=', self.env.context['accrual_entry_date'])]
         return domain
 
+    @api.model
+    def _get_delivered_quantity_by_analytic_domain(self):
+        """"To make sure we don't consider timesheets while computing delivered quantities
+        from non timesheet AALs (e.g. Services and Materials)"""
+        domain = super()._get_delivered_quantity_by_analytic_domain()
+        return Domain.AND([domain, Domain('project_id', '=', False)])
+
     ###########################################
     # Service : Project and task generation
     ###########################################
@@ -163,9 +162,9 @@ class SaleOrderLine(models.Model):
         lines_by_timesheet = self.filtered(lambda sol: sol.product_id and sol.product_id._is_delivered_timesheet())
         domain = Domain(lines_by_timesheet._timesheet_compute_delivered_quantity_domain())
         refund_account_moves = self.order_id.invoice_ids.filtered(lambda am: am.state == 'posted' and am.move_type == 'out_refund').reversed_entry_id
-        timesheet_domain = Domain('timesheet_invoice_id', '=', False) | Domain('timesheet_invoice_id.state', '=', 'cancel')
+        timesheet_domain = Domain('reinvoice_move_id', '=', False) | Domain('reinvoice_move_id.state', '=', 'cancel')
         if refund_account_moves:
-            credited_timesheet_domain = Domain('timesheet_invoice_id.state', '=', 'posted') & Domain('timesheet_invoice_id', 'in', refund_account_moves.ids)
+            credited_timesheet_domain = Domain('reinvoice_move_id.state', '=', 'posted') & Domain('reinvoice_move_id', 'in', refund_account_moves.ids)
             timesheet_domain |= credited_timesheet_domain
         domain &= timesheet_domain
         if start_date:
@@ -205,3 +204,14 @@ class SaleOrderLine(models.Model):
     @api.model
     def _get_product_service_policy(self):
         return super()._get_product_service_policy() + ['delivered_timesheet']
+
+    def _is_line_reinvoicable(self):
+        """If line is supposed to be delivered based on a timesheet then we skip the expense policy
+        check in super."""
+        if self.product_id.service_type == 'timesheet':
+            return (
+                not self.is_expense
+                and self.product_id.invoice_policy == 'delivery'
+            )
+
+        return super()._is_line_reinvoicable()
